cmd/create/addon/cmd.go

```diff
@@ -0,0 +1,150 @@
+	Run: run,
+}
+
+func init() {
+	flags := Cmd.Flags()
+
+	flags.StringVarP(
+		&args.clusterKey,
+		"cluster",
+		"c",
+		"",
+		"Name or ID of the cluster to add the IdP to (required).",
+	)
+	Cmd.MarkFlagRequired("cluster")
+}
+
+func run(_ *cobra.Command, argv []string) {
+	reporter := rprtr.CreateReporterOrExit()
+	logger := logging.CreateLoggerOrExit(reporter)
+
+	// Check command line arguments:
+	if len(argv) != 1 {
+		reporter.Errorf("Expected exactly one command line parameters containing the identifier of the add-on.")
+		os.Exit(1)
+	}
+
+	addOnID := argv[0]
+	if addOnID == "" {
+		reporter.Errorf("Add-on ID is required.")
+		os.Exit(1)
+	}
+
+	// Check that the cluster key (name, identifier or external identifier) given by the user
+	// is reasonably safe so that there is no risk of SQL injection:
+	clusterKey := args.clusterKey
+	if !ocm.IsValidClusterKey(clusterKey) {
+		reporter.Errorf(
+			"Cluster name, identifier or external identifier '%s' isn't valid: it "+
+				"must contain only letters, digits, dashes and underscores",
+			clusterKey,
+		)
+		os.Exit(1)
+	}
+
+	// Create the AWS client:
+	awsClient, err := aws.NewClient().
+		Logger(logger).
+		Build()
+	if err != nil {
+		reporter.Errorf("Failed to create AWS client: %v", err)
+		os.Exit(1)
+	}
+
+	awsCreator, err := awsClient.GetCreator()
+	if err != nil {
+		reporter.Errorf("Failed to get AWS creator: %v", err)
+		os.Exit(1)
+	}
+
+	// Create the client for the OCM API:
+	ocmConnection, err := ocm.NewConnection().
+		Logger(logger).
+		Build()
+	if err != nil {
+		reporter.Errorf("Failed to create OCM connection: %v", err)
+		os.Exit(1)
+	}
+	defer func() {
+		err = ocmConnection.Close()
+		if err != nil {
+			reporter.Errorf("Failed to close OCM connection: %v", err)
+		}
+	}()
+
+	// Get the client for the OCM collection of clusters:
+	clustersCollection := ocmConnection.ClustersMgmt().V1().Clusters()
+
+	// Try to find the cluster:
+	reporter.Debugf("Loading cluster '%s'", clusterKey)
+	cluster, err := ocm.GetCluster(clustersCollection, clusterKey, awsCreator.ARN)
+	if err != nil {
+		reporter.Errorf("Failed to get cluster '%s': %v", clusterKey, err)
+		os.Exit(1)
+	}
+
+	if cluster.State() != cmv1.ClusterStateReady {
+		reporter.Errorf("Cluster '%s' is not yet ready", clusterKey)
+		os.Exit(1)
+	}
+
+	addOnInstallation, err := cmv1.NewAddOnInstallation().
+		Addon(cmv1.NewAddOn().ID(addOnID)).
+		Build()
+	if err != nil {
+		reporter.Errorf("Failed to create add-on installation '%s' for cluster '%s'", addOnID, clusterKey)
+		os.Exit(1)
+	}
+	reporter.Debugf("Installing add-on '%s' on cluster '%s'", addOnID, clusterKey)
+	_, err = clustersCollection.Cluster(cluster.ID()).
+		Addons().
+		Add().
+		Body(addOnInstallation).
+		Send()
+	if err != nil {
+		reporter.Errorf("Failed to add add-on installation '%s' for cluster '%s': %s", addOnID, clusterKey, err)
+		os.Exit(1)
+	}
+	reporter.Debugf("Installed add-on '%s' on cluster '%s'", addOnID, clusterKey)
```

Comment:
Could you pop this into something reusable in pkg? osde2e could definitely use this.

Reply:
Ah, good point - I'll add to the cluster provider

Reply:
How about now? I've added `clusterprovider.InstallAddOn()`, which I think you should be able to use straight up in osde2e without having to wrap around OCM

Reply:
Would it be possible to take the aws related code and push this inside that function as well? That way it more encapsulates the extra functionality that moactl is providing.

If not, we can work around, but I like the AWS related code sitting in the function because that's a lot of the primary value of moactl, I think.

Reply:
I think what you're describing is better accomplished in separate refactor commit. I'm currently following the same pattern as clusterprovider.GetCluster, UpdateCluster, DeleteCluster, etc. They all take the clusterKey and the ARN. The only exception is creating the cluster, which needs it for grabbing the key/secret pair and tagging the IAM user.

I'd be happy to have an offline chat to discuss a refactor that would make it easier to consume externally. What do you think?

Reply:
+1 to that, that makes sense. @jharrington22 has mentioned this in the past as well, but I totally agree. No need to discuss offline (especially since I'm going on PTO next week :-)) Maybe that's something we can take on. I'll file a JIRA for this against SDCICD to help refactor that bit.

Reply:
Sounds good, thanks! :) Enjoy your time off!